### PR TITLE
add cf flags to 2D masks

### LINKED ISSUE #346: add `flag_values` and `flag_meanings` to 2D masks?
This would allow to indicate which region corresponds to which number via `attrs` and would work nicely with `cf_xarray`.

```python
import cf_xarray
import xarray as xr

da = xr.DataArray(
    [[1, 2, 1], [1, 2, np.NaN]],
    dims=("lat", "lon"),
    attrs={
        "flag_values": [1, 2],
        "flag_meanings": "atlantic_ocean pacific_ocean",
        "standard_name": "region",
    },
  name="mask",
)

print(da.cf)

print(da.cf == "atlantic_ocean")

print(da.cf.isin(["indian_ocean", "pacific_ocean"]))

```

See
- https://cf-xarray.readthedocs.io/en/latest/flags.html
- xarray-contrib/cf-xarray#199
- http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#flags

Open questions:
- do we add `abbrevs` or `names` (probably `abbrevs`)
- does it work for `float` arrays (as I encode "not belonging to any region" with `NaN`)?

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #346
 - [x] Tests added
 - [x] Passes `isort . && black . && flake8`
 - [x] Fully documented, including `whats-new.rst`

A WIP to add cf-compliant flags.

---

Example:

```python
import cf_xarray
import numpy as np
import regionmask

lon = np.arange(0.5, 360, 1)
lat = np.arange(89.5, -90, -1)

mask = regionmask.defined_regions.ar6.land.mask(lon, lat)

# leverage cf_xarray to select a region
mask_WCE = mask.cf == "WCE"

# plot
ax = regionmask.defined_regions.ar6.land.plot(add_label=False)
mask_WCE.plot(ax=ax, colors=["none", "#99d8c9"], levels=[0.95, 1.05], add_colorbar=False)
plt.savefig("flags.png", dpi=300, bbox_inches="tight")
```

flags

The mask would then look something like this:

```
<xarray.DataArray 'mask' (lat: 180, lon: 360)>
array([[...]])
Coordinates:
  * lat      (lat) float64 89.5 88.5 87.5 86.5 85.5 ... -86.5 -87.5 -88.5 -89.5
  * lon      (lon) float64 0.5 1.5 2.5 3.5 4.5 ... 355.5 356.5 357.5 358.5 359.5
Attributes:
    standard_name:  region
    flag_values:    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16...
    flag_meanings:  GIC NWN NEN WNA CNA ENA NCA SCA CAR NWS NSA NES SAM SWS S...
```

### CODE REVIEW

File: docs/notebooks/mask_2D.ipynb [hunk 1]
```diff
@@ -276,7 +276,89 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mask_names = regionmask.defined_regions.srex.mask(airtemps, flag=\"names\")\n",
+    "\n",
+    "try:\n",
+    "    mask_names.cf == \"C. North_America\"\n",
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:343]:
```suggestion
    "    mask_names.cf == \"C. North America\"\n",
```

File: docs/notebooks/mask_2D.ipynb [hunk 2]
```diff
@@ -276,7 +276,89 @@
...
+    "except ValueError:\n",
+    "    print(\"error raised\")\n",
+    "\n",
+    "mask_names.cf == \"C. North_America\".replace(\" \", \"_\")"
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:347]:
```suggestion
    "mask_names.cf == \"C. North America\".replace(\" \", \"_\")"
```

File: docs/notebooks/mask_2D.ipynb [hunk 3]
```diff
@@ -276,7 +276,89 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Select a specific region - traditional way"
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:354]:
```suggestion
    "### Select a specific region - without cf_xarray"
```

File: docs/notebooks/mask_2D.ipynb [hunk 4]
```diff
@@ -276,7 +276,89 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Without the help of xf_xarray we first need to find out which number the reigion 'C. North America' corresponds to:"
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:361]:
```suggestion
    "We first need to find out which number the region 'C. North America' corresponds to:"
```

File: docs/notebooks/mask_2D.ipynb [hunk 5]
```diff
@@ -300,7 +391,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "`xarray` provides the handy `where` method ([documentation](https://xarray.pydata.org/en/stable/generated/xarray.DataArray.where.html)), that filters elements from this object according to a condition:"
+    "To replace all grid points outside of CNA we use the `where` method of `xr.Dataset` ([documentation](https://xarray.pydata.org/en/stable/generated/xarray.DataArray.where.html)), which filters elements from this object according to a condition:"
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:394]:
```suggestion
    "To replace all grid points outside of CNA we use the `where` method of `xr.Dataset` ([documentation](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.where.html)), which filters elements from this object according to a condition:"
```

File: regionmask/core/mask.py [hunk 6]
```diff
@@ -97,6 +98,15 @@
     There is (currently) no automatic detection of overlapping regions.
 """
 
+_FLAG_DOCSTRING = """\
+flag : str, default "abbrevs"
+    Indicates if the "abbrevs" (abbreviations) or "names" should be added as
+    `flag_values` and `flag_meanings` to the attributes (`attrs`) of the mask. If None
+    nothing is added. Using cf_xarray these can be used to select single
+    (``mask.cf == "CNA"``) or multiple (``mask.cf.isin``) regions. Note that spaces are
+    replaced by underscore.
```
Comment (mathause) [inline regionmask/core/mask.py:107]:
```suggestion
    replaced by underscores.
```

File: docs/notebooks/mask_2D.ipynb [hunk 7]
```diff
@@ -573,9 +664,9 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 (ipykernel)",
+   "display_name": "Python [conda env:regionmask-docs]",
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:667]:
```suggestion
   "display_name": "Python 3 (ipykernel)",
```

File: docs/notebooks/mask_2D.ipynb [hunk 8]
```diff
@@ -573,9 +664,9 @@
...
    "language": "python",
-   "name": "python3"
+   "name": "conda-env-regionmask-docs-py"
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:669]:
```suggestion
   "name": "python3"
```

File: docs/notebooks/mask_2D.ipynb [hunk 9]
```diff
@@ -587,7 +678,12 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.10.2"
+   "version": "3.9.13"
+  },
+  "vscode": {
+   "interpreter": {
+    "hash": "45192e058903e72007b0902b5a48c99e46b48f33e794873f2cde694207f1593d"
+   }
```
Comment (mathause) [inline docs/notebooks/mask_2D.ipynb:686]:
```suggestion
   "version": "3.10.2"
```

**Comment by jbusecke** [comment]
Still have this on my list to test, will try to get to it tomorrow. Sorry for the delay.

**Comment by mathause** [comment]
No worries - another look is certainly very welcome.

That being said I went over the code again and from my side this is ready to be merged.

**Comment by jbusecke** [comment]
Oh sorry, I went on vacation and was kind of swamped before. I will try to work this into the xMIP docs soon and then report back if I find anything. Thanks a lot for working on this!

**Pull request merged**